### PR TITLE
Import matches

### PR DESCRIPTION
**Pull request by bartoszmaka**

### CODE REVIEW

File: app/services/import_matches.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    def home_team
+      Team.find_by!(name: match_attributes.home_team_name)
+    end
+
+    def away_team
+      Team.find_by(name: match_attributes.away_team_name)
+    end
+
+    def update_football_score
+      football_match
```
Comment (stevo) [inline app/services/import_matches.rb]: obsolete line

File: app/services/import_matches.rb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        round_id: round.id
+      )
+    end
+
+    def home_team
+      Team.find_by!(name: match_attributes.home_team_name)
+    end
+
+    def away_team
+      Team.find_by(name: match_attributes.away_team_name)
```
Comment (stevo) [inline app/services/import_matches.rb]: `!`

File: app/services/import_matches.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  class UpsertMatch < Patterns::Service
+    attr_reader :round, :match_attributes
+
+    def initialize(round:, match_attributes:)
+      @round = round
+      @match_attributes = match_attributes
+    end
+
+    def call
+      update_football_score
```
Comment (stevo) [inline app/services/import_matches.rb]: drop using `football` in names here as we are not differentiating between football and other matches

File: spec/features/root_page_spec.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+
+feature 'Home Page' do
+  scenario 'visit root page ' do
+    visit root_path
+    expect(page).to have_table_row(
```
Comment (stevo) [inline spec/features/root_page_spec.rb]: I think you can start writing real expectations now :)

File: spec/support/matchers/have_table_row.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
+require 'rspec/expectations'
+require 'pry'
+
+RSpec::Matchers.define :have_table_row do |*expected|
+  match do |page|
```
Comment (stevo) [inline spec/support/matchers/have_table_row.rb]: we will refactor it tomorrow

File: spec/features/root_page_spec.rb [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    monaco    = create(:team, name: 'Monaco')
+
+    round = Round.create(number: 1, year: 2017)
+
+    create(
+      :football_match,
+      match_date: DateTime.new(2017, 9, 20, 12, 30),
+      home_team: barcelona,
+      home_team_score: 0,
+      away_team_score: 22,
```
Comment (stevo) [inline spec/features/root_page_spec.rb]: how about score like  `2 : 5` ?

File: spec/features/root_page_spec.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      :football_match,
+      match_date: DateTime.new(2017, 9, 21, 12, 30),
+      home_team: juventus,
+      home_team_score: nil,
+      away_team_score: nil,
+      away_team: monaco,
+      round_id: round.id
+    )
+
+    Timecop.freeze(DateTime.new(2017, 9, 20, 18, 30))
```
Comment (stevo) [inline spec/features/root_page_spec.rb]: use `travel_to {}` and remove `Timecop` gem

File: spec/features/root_page_spec.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    visit root_path
+
+    expect(page).to have_table_row(
```
Comment (stevo) [inline spec/features/root_page_spec.rb]: why do you use two different syntaxes for matcher ?

File: spec/features/root_page_spec.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+      'Match Date' => '2017-09-20 14:30:00',
```
Comment (stevo) [inline spec/features/root_page_spec.rb]: do we really need to know exact second ?

File: spec/features/root_page_spec.rb [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+      'Home Team' => 'Barcelona',
+      'Away Team' => 'Real Madrid',
+      'Score Home' => '0',
+      'Score Away' => '22'
+    )
+    expect(page).to have_table_row('2017-09-20 14:30:00', 'Barcelona', '0', include('22'), 'Real Madrid')
```
Comment (stevo) [inline spec/features/root_page_spec.rb]: can we change `'0'` to `0` ? And what is the reason for `include` ?

File: spec/features/root_page_spec.rb [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    expect(page).to have_table_row('2017-09-21 14:30:00', 'Juventus', '0', '0', 'Monaco', include('Place Bet'))
+    expect(page.all('tbody > tr').first).not_to have_link('Place Bet')
```
Comment (stevo) [inline spec/features/root_page_spec.rb]: why ?

File: app/views/football_matches/index.html.erb [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+      <th>Home Team</th>
+      <th>Score Home</th>
+      <th>Score Away</th>
+      <th>Away Team</th>
+      <th></th>
+    </tr>
+  </thead>
+
+  <tbody>
+    <% @football_matches.each do |football_match| %>
```
Comment (stevo) [inline app/views/football_matches/index.html.erb]: use `render locals: {}` in controller and drop `@` here

File: app/views/football_matches/index.html.erb [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+      <tr>
+        <td><%= football_match.match_date.localtime.to_formatted_s(:db) %></td>
+        <td><%= football_match.home_team.name %></td>
+        <td><%= football_match.home_team_score || 0 %></td>
+        <td><%= football_match.away_team_score || 0 %></td>
+        <td><%= football_match.away_team.name %></td>
+        <td><%= link_to "Place Bet" if football_match.match_date > 1.hour.ago %></td>
```
Comment (stevo) [inline app/views/football_matches/index.html.erb]: where can I find specs for this ?

**Review by stevo — CHANGES_REQUESTED** [review]

File: app/services/import_matches.rb [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    def home_team
+      Team.find_by!(name: match_attributes.home_team_name)
+    end
+
+    def away_team
+      Team.find_by!(name: match_attributes.away_team_name)
+    end
+
+    def update_score
+      match
```
Comment (stevo) [inline app/services/import_matches.rb]: why do we need this line ?

File: spec/factories/football_match.rb [hunk 15]
```diff
@@ -0,0 +1,9 @@
+FactoryGirl.define do
+  factory :football_match do
+    match_date DateTime.parse('20.09.2013')
```
Comment (stevo) [inline spec/factories/football_match.rb]: `DateTime.new`

File: spec/factories/football_match.rb [hunk 16]
```diff
@@ -0,0 +1,9 @@
...
+    home_team 'Barcelona'
+    away_team 'Real Madrid'
+    home_team_score 0
```
Comment (stevo) [inline spec/factories/football_match.rb]: use traits to differ between `completed` and `pending` matches

File: spec/factories/football_match.rb [hunk 17]
```diff
@@ -0,0 +1,9 @@
+FactoryGirl.define do
+  factory :football_match do
+    match_date DateTime.parse('20.09.2013')
+    home_team 'Barcelona'
```
Comment (stevo) [inline spec/factories/football_match.rb]: Is it string, or an object ?

File: spec/features/root_page_spec.rb [hunk 18]
```diff
@@ -0,0 +1,51 @@
+require 'rails_helper'
+
+feature 'Home Page' do
+  scenario 'lists all matches with Place Bet button' do
+    barcelona = create(:team, name: 'Barcelona')
+    real      = create(:team, name: 'Real Madrid')
+    juventus  = create(:team, name: 'Juventus')
+    monaco    = create(:team, name: 'Monaco')
+
+    round = Round.create(number: 1, year: 2017)
```
Comment (stevo) [inline spec/features/root_page_spec.rb]: use factories

File: spec/features/root_page_spec.rb [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+    )
+
+    expect(page).to have_table_row(second_match.match_date.localtime.to_formatted_s(:db), 'Juventus', '0', '0', 'Monaco', include('Place Bet'))
+    expect(page.all('tbody > tr').first).not_to have_link('Place Bet')
+    expect(page.all('tbody > tr').last).to have_link('Place Bet')
+
+    travel_to DateTime.new(2017, 9, 22, 12, 30)
+    refresh
+
+    expect(page.all('tbody > tr').last).not_to have_link('Place Bet')
```
Comment (stevo) [inline spec/features/root_page_spec.rb]: after refresh the link is gone... why ? How should I understand that ?

File: spec/features/root_page_spec.rb [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+
+    expect(page).to have_table_row(
+      'Match Date' => first_match.match_date.localtime.to_formatted_s(:db),
+      'Home Team' => 'Barcelona',
+      'Away Team' => 'Real Madrid',
+      'Score Home' => '2',
+      'Score Away' => '5'
+    )
+
+    expect(page).to have_table_row(second_match.match_date.localtime.to_formatted_s(:db), 'Juventus', '0', '0', 'Monaco', include('Place Bet'))
```
Comment (stevo) [inline spec/features/root_page_spec.rb]: `expect(1+2*5/10*20-1922).to eq (1+2*5/10*20-1922)`....

File: spec/features/root_page_spec.rb [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+      home_team_score: nil,
+      away_team_score: nil,
+      away_team: monaco,
+      round_id: round.id
+    )
+
+    travel_to DateTime.new(2017, 9, 20, 18, 30)
+    visit root_path
+
+    expect(page).to have_table_row(
```
Comment (stevo) [inline spec/features/root_page_spec.rb]: any reason for using keys here and not using them in assertion below ?

File: spec/matchers/have_table_row_spec.rb [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+        HTML
+
+        within(page.find('table#second-table')) do
+          expect(page).to have_table_row('John', 27)
+        end
+      end
+    end
+  end
+end
+# expect(page).to have_table_row(0 => "John", 1 => 27)
```
Comment (stevo) [inline spec/matchers/have_table_row_spec.rb]: ✂️

**Review by stevo — CHANGES_REQUESTED** [review]

File: .travis.yml [hunk 23]
```diff
@@ -1,6 +1,6 @@
 rvm:
   - 2.4.0
-bundler_args: --without development
+# bundler_args: --without development
```
Comment (stevo) [inline .travis.yml]: Why ?

File: app/services/import_teams.rb [hunk 24]
```diff
@@ -0,0 +1,15 @@
+require 'open-uri'
+class ImportTeams < Patterns::Service
+  URL = 'http://www.goal.com/en/uefa-champions-league/fixtures-results/week-1/4oogyu6o156iphvdvphwpck10'.freeze
```
Comment (stevo) [inline app/services/import_teams.rb]: This should go to ENV

File: app/views/football_matches/index.html.erb [hunk 25]
```diff
@@ -0,0 +1,26 @@
+
```
Comment (stevo) [inline app/views/football_matches/index.html.erb]: ✂️

File: app/views/football_matches/index.html.erb [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+  <tbody>
+    <% football_matches.each do |football_match| %>
+      <tr>
+        <td><%= l(football_match.match_date, format: :date_with_time) %></td>
+        <td><%= football_match.home_team.name %></td>
+        <td><%= football_match.home_team_score || 0 %></td>
+        <td><%= football_match.away_team_score || 0 %></td>
+        <td><%= football_match.away_team.name %></td>
+        <td><%= link_to "Place Bet" if football_match.match_date > (DateTime.now + 1.hour) %></td>
```
Comment (stevo) [inline app/views/football_matches/index.html.erb]: I though we discussed moving `football_match.match_date > (DateTime.now + 1.hour)` out of this file ?

File: db/migrate/20170915135145_add_round_to_football_matches.rb [hunk 27]
```diff
@@ -0,0 +1,6 @@
+class AddRoundToFootballMatches < ActiveRecord::Migration[5.1]
+  def change
+    add_column :football_matches, :round_id, :integer
+    add_foreign_key :football_matches, :rounds
```
Comment (stevo) [inline db/migrate/20170915135145_add_round_to_football_matches.rb]: what is the difference here (between this and line above?) do we need both?

File: spec/factories/team.rb [hunk 28]
```diff
@@ -0,0 +1,5 @@
+FactoryGirl.define do
+  factory :team do
+    sequence :name { |x| "FactoryTeam#{x}" }
```
Comment (stevo) [inline spec/factories/team.rb]: "FactoryTeam" ? :P

File: spec/rails_helper.rb [hunk 29]
```diff
@@ -29,6 +31,18 @@
...
+    DatabaseCleaner.strategy = :transaction
+    DatabaseCleaner.clean_with(:truncation)
+  end
+
+  config.around(:each) do |example|
+    DatabaseCleaner.cleaning do
+      example.run
+    end
+  end
+  config.include FactoryGirl::Syntax::Methods
```
Comment (stevo) [inline spec/rails_helper.rb]: keep includes to the top of this block

File: spec/support/capybara.rb [hunk 30]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (stevo) [inline spec/support/capybara.rb]: ✂️

File: spec/support/capybara.rb [hunk 31]
```diff
@@ -0,0 +1,5 @@
...
+require_relative 'locators/find_table_row'
+
+Capybara::Session.include FindTableRow
```
Comment (stevo) [inline spec/support/capybara.rb]: can we remove this line ?

File: spec/support/matchers/have_table_row.rb [hunk 32]
```diff
@@ -0,0 +1,17 @@
+require 'rspec/expectations'
+require 'pry'
```
Comment (stevo) [inline spec/support/matchers/have_table_row.rb]: we probably do not need that ;)

File: spec/vcr_helper.rb [hunk 33]
```diff
@@ -0,0 +1,7 @@
+require 'rubygems'
```
Comment (stevo) [inline spec/vcr_helper.rb]: ?

**Review by stevo — CHANGES_REQUESTED** [review]
Couple of small things

File: spec/support/capybara.rb (hunk 31 above)
Comment (synion) [reply spec/support/capybara.rb]: we checked this but we don't have access to this method without this line. In feature spec method have_table_row does not work

**Review by stevo — APPROVED** [review]
Ok to merge 🍹

**Pull request merged**